rpc_util.go

```diff
@@ -160,46 +160,82 @@ type EmptyCallOption struct{}
 func (EmptyCallOption) before(*callInfo) error { return nil }
 func (EmptyCallOption) after(*callInfo)        {}
 
-type beforeCall func(c *callInfo) error
-
-func (o beforeCall) before(c *callInfo) error { return o(c) }
-func (o beforeCall) after(c *callInfo)        {}
-
-type afterCall func(c *callInfo)
-
-func (o afterCall) before(c *callInfo) error { return nil }
-func (o afterCall) after(c *callInfo)        { o(c) }
-
 // Header returns a CallOptions that retrieves the header metadata
 // for a unary RPC.
 func Header(md *metadata.MD) CallOption {
-	return afterCall(func(c *callInfo) {
-		if c.stream != nil {
-			*md, _ = c.stream.Header()
-		}
-	})
+	return HeaderCallOption{md: md}
+}
+
+// HeaderCallOption is a CallOption for collecting response header metadata.
+// This is an EXPERIMENTAL API.
+type HeaderCallOption struct {
+	md *metadata.MD
+}
+
+func (o HeaderCallOption) before(c *callInfo) error { return nil }
+func (o HeaderCallOption) after(c *callInfo) {
+	if c.stream != nil {
+		*o.md, _ = c.stream.Header()
+	}
+}
+
+// HeaderAddr returns the address where header metadata will be stored after the
```

Comment:
I'd prefer to export the field vs. add an accessor.  It would be simpler to use, and these CallOption types are passed around by value anyway, so we don't have to worry about the interceptor modifying something inside the application (except the header/trailer/etc data directly).  Would you have any concerns with that?

Reply:
That totally makes sense. The methods only made sense when relying on caller to do an interface-type-assertion to call it. With exported types, this is much better.